iceoryx_hoofs/include/iceoryx_hoofs/error_reporting/api.hpp

```diff
@@ -0,0 +1,125 @@
+#ifndef IOX_HOOFS_ERROR_REPORTING_API_HPP
+#define IOX_HOOFS_ERROR_REPORTING_API_HPP
+
+#include "error_forwarding.hpp"
+#include "error_kind.hpp"
+#include "platform/error_reporting.hpp"
+
+// clang-format on
+
+/// @brief transforms an error code to an error object according to
+///        default specification or override by module
+/// @code error code to be transformed
+/// @note this relies on overloading
+#define IOX_ERROR(code) iox::err::toError(code)
+
+// The following macros are statements (not expressions).
+// This is important, as it enforces correct use to some degree.
+// For example thye cannot be used as function arguments and must be terminated with a ';'.
+
+/// @brief calls panic handler and does not return
+/// @param msg optional message string literal
+/// @note could actually throw if desired without breaking control flow asssumptions
+#define IOX_PANIC(...)                                                                                                 \
+    do                                                                                                                 \
+    {                                                                                                                  \
+        iox::err::panic(CURRENT_SOURCE_LOCATION);                                                                      \
```

Comment:
Here `panic` from the platform is directly called, right? If so, one could introduce a `panic` function which doesn't abort execution. Shouldn't we therefore call `forwardError` with `FATAL`?

Reply:
This is basically the question whether we want to expose `IOX_PANIC` to the user.

`IOX_PANIC` is lower level, it does not need a reason/error on its own to forward, `forwardError` with `FATAL` does.

Reply:
Good point, must still fix it.

Reply:
Fixed. Now error_forwarding.hpp takes care of all of the noreturn semantics. (unless something is missing)